### PR TITLE
Feature: Implement the fundamental for rich progress

### PR DESCRIPTION
**Pull request by hez2010**
**Resolved / Related Issues**
Items resolved / related issues by this PR.
- Prerequisites of #10755, #8821

In this PR, I did two things:
- Merge all `IProgress<float>` and `IProgress<FileSystemStatusCode>` to `IProgress<FileSystemProgress>`
- Introduce a wrapper type `FileSystemProgress` of `IProgress<FileSystemProgress>` to maintain progress information without needs of maintaining those information in the each implementation of `FileSystemOperation`

This PR didn't change how we report the progress, with this fundamental we can improve them one by one in the future.

An example of zip extraction progress:

image

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [ ] Tested the changes for accessibility

**Screenshots (optional)**
Add screenshots here.

### CODE REVIEW

File: src/Files.App/Filesystem/Archive/ArchiveCreator.cs [hunk 1]
```diff
@@ -118,6 +130,16 @@ public async Task<bool> RunCreationAsync()
...
+			fsProgress.Percentage = null;
+            fsProgress.ReportStatus(Shared.Enums.FileSystemStatusCode.Success);
+        }
+
+        private void Compressor_Compressing(object? _, ProgressEventArgs e)
+		{
+			fsProgress.Percentage = e.PercentDone;
+			fsProgress.Report();
+        }
 	}
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/Archive/ArchiveCreator.cs:144]: The overall formatting of this file is broken. Could you make pass VS formatter on it while it's being modified?

File: src/Files.App/Filesystem/FilesystemOperations/FileSystemProgress.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                {
+                    criticalReport = true;
+                }
+                enumerationCompleted = value;
+            }
+        }
+        /// <summary>
+        /// Only used when detailed count isn't available.
+        /// </summary>
+        public float? Percentage { get; set; }
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/FilesystemOperations/FileSystemProgress.cs:49]: Could it be worth to wrap both the Percentage and the Count inside an object that would adapt its behavior depending on the data avalaible to it?

File: src/Files.App/Filesystem/FilesystemOperations/FileSystemProgress.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            this.sampler = new(samplerInterval);
+        }
+
+        public void Report(float? percentage = null)
+        {
+            Percentage = percentage;
+            if ((EnumerationCompleted && (ProcessedItemsCount == ItemsCount || ProcessedSize == TotalSize)) ||
+                (percentage is float f && MathF.Abs(f - 100f) <= float.Epsilon))
+            {
+                if (status is FileSystemStatusCode.InProgress or null)
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/FilesystemOperations/FileSystemProgress.cs:64]: This conditions are a mouthful to understand and the two `if`s can be merged together.
Worth putting it inside a method.

File: src/Files.App/Filesystem/FilesystemOperations/FileSystemProgress.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                {
+                    status = FileSystemStatusCode.Success;
+                    CompletedTime = DateTimeOffset.Now;
+                }
+            }
+            if (status is FileSystemStatusCode.Success) CompletedTime = DateTimeOffset.Now;
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/FilesystemOperations/FileSystemProgress.cs:70]: `CompletedTime = DateTimeOffset.Now;` should be on another line.

File: src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs [hunk 5]
```diff
@@ -46,17 +46,25 @@ public FilesystemOperations(IShellPage associatedInstance)
...
+            fsProgress.Report();
+
+            try
 			{
 				switch (source.ItemType)
 				{
 					case FilesystemItemType.File:
 						{
-							var newEntryInfo = await ShellNewEntryExtensions.GetNewContextMenuEntryForType(Path.GetExtension(source.Path));
+                            var newEntryInfo = await ShellNewEntryExtensions.GetNewContextMenuEntryForType(Path.GetExtension(source.Path));
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs:67]: Formatting

File: src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs [hunk 6]
```diff
@@ -46,17 +46,25 @@ public FilesystemOperations(IShellPage associatedInstance)
...
-			try
+            FileSystemProgress fsProgress = new(progress)
+            {
+                ItemsCount = 1,
+                EnumerationCompleted = true,
+				Status = FileSystemStatusCode.InProgress
+            };
+            fsProgress.Report();
+
+            try
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs:61]: Formatting

File: src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs [hunk 7]
```diff
@@ -90,8 +98,8 @@ public async Task<(IStorageHistory, IStorageItem)> CreateAsync(IStorageItemWithP
 						}
 
 					case FilesystemItemType.Directory:
-						{
-							var fsFolderResult = await associatedInstance.FilesystemViewModel.GetFolderFromPathAsync(PathNormalization.GetParentDir(source.Path));
+                        {
+                            var fsFolderResult = await associatedInstance.FilesystemViewModel.GetFolderFromPathAsync(PathNormalization.GetParentDir(source.Path));
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs:102]: Formatting

File: src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs [hunk 8]
```diff
@@ -111,37 +119,42 @@ public async Task<(IStorageHistory, IStorageItem)> CreateAsync(IStorageItemWithP
 						break;
 				}
 
-				errorCode?.Report(fsResult);
+                fsProgress.ProcessedItemsCount = 1;
+                fsProgress.ReportStatus(fsResult);
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs:123]: This file could use a full reformatting by VS, I won't mark any other issue of formatting on it therefore ^^

File: src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs [hunk 9]
```diff
@@ -880,14 +912,22 @@ public async Task<IStorageHistory> CopyItemsAsync(IList<IStorageItemWithPath> so
...
-			var rawStorageHistory = new List<IStorageHistory>();
+		public async Task<IStorageHistory> MoveItemsAsync(IList<IStorageItemWithPath> source, IList<string> destination, IList<FileNameConflictResolveOptionType> collisions, IProgress<FileSystemProgress> progress, CancellationToken token)
+        {
+            FileSystemProgress fsProgress = new(progress)
+            {
+                ItemsCount = source.Count,
+                EnumerationCompleted = true,
+                Status = FileSystemStatusCode.InProgress
+            };
+            fsProgress.Report();
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs:928]: This bit of code is used a lot in through this file, could it be worth it to put it inside a builder method initializing the whole process and returning it?

File: src/Files.App/ViewModels/StatusCenterViewModel.cs [hunk 10]
```diff
@@ -187,51 +187,89 @@ public PostedStatusBanner(StatusBanner banner, IOngoingTasksActions OngoingTasks
...
-		private void ReportProgressToBanner(float value)
+		private void ReportProgressToBanner(FileSystemProgress value)
 		{
 			if (CancellationToken.IsCancellationRequested) // file operation has been cancelled, so don't update the progress text
 			{
 				return;
 			}
 
-			if (value <= 100.0f)
+			if (value.Status is FileSystemStatusCode status) Banner.Status = status.ToStatus();
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/StatusCenterViewModel.cs:220]: `Banner.Status = status.ToStatus();` needs to be on a new line.

**Review by QuaintMako — COMMENTED** [review]
It looks very promising! Good work on what has been done already!
On a sidenote, I stopped noting it after a while, but all the files modified need to be reformated to be consistant with the existing codebase.

File: src/Files.App/Filesystem/Archive/ArchiveCreator.cs (hunk 1 above)
Comment (hez2010) [reply src/Files.App/Filesystem/Archive/ArchiveCreator.cs:144]: It was broken due to the difference of spaces and tabs. My Visual Studio keeps replacing tabs to spaces wherever I have edited.

File: src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs (hunk 9 above)
Comment (hez2010) [reply src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs]: I decided to put them in the constructor.

File: src/Files.App/Filesystem/FilesystemOperations/FileSystemProgress.cs (hunk 2 above)
Comment (hez2010) [reply src/Files.App/Filesystem/FilesystemOperations/FileSystemProgress.cs]: Percentage here effectively overrides the details due to count not available (particularly for `ShellFilesOperation`). I prefer not to wrap them inside an object to keep it simple.

**Comment by hez2010** [comment]
I have run `dotnet format` to fix all indent related issues.
PTAL.

File: .editorconfig [hunk 11]
```diff
@@ -0,0 +1,4 @@
+[*.cs]
+indent_style = tab
+indent_size = 4
+tab_width = 4
```
Comment (yaira2) [inline .editorconfig:4]: I like this!

**Review by QuaintMako — APPROVED** [review]
LGTM.

**Review by yaira2 — APPROVED** [review]
LGTM

**Pull request merged**